Mathlib/Analysis/Normed/Ring/FiniteExtension.lean

```diff
@@ -0,0 +1,251 @@
+  classical
+  simp only [norm, repr_algebraMap _ hBi]
+  apply le_antisymm
+  · aesop
+  · exact le_sup'_of_le _ (mem_univ i) (by simp)
+
+/-- For any `K`-basis of `L`, if the norm on `K` is nonarchimedean, then so is `B.norm`. -/
+theorem norm_isNonarchimedean (hna : IsNonarchimedean (Norm.norm : K → ℝ)) :
+    IsNonarchimedean B.norm := fun x y ↦ by
+  obtain ⟨ixy, _, hixy⟩ := exists_mem_eq_sup' univ_nonempty (fun i ↦ ‖(B.repr (x + y)) i‖)
+  have hxy : ‖B.repr (x + y) ixy‖ ≤ max ‖B.repr x ixy‖ ‖B.repr y ixy‖ := by
+    rw [LinearEquiv.map_add, Finsupp.coe_add, Pi.add_apply]; exact hna _ _
+  rw [Basis.norm, hixy]
+  rcases le_max_iff.mp hxy with (hx | hy)
+  · exact le_max_of_le_left (le_trans hx (norm_repr_le_norm B ixy))
+  · exact le_max_of_le_right (le_trans hy (norm_repr_le_norm B ixy))
+
+/-- For any `K`-basis of `L`, `B.norm` is bounded with respect to multiplication. That is,
+  `∃ (c : ℝ), c > 0` such that ` ∀ (x y : L), B.norm (x * y) ≤ c * B.norm x * B.norm y`. -/
+theorem norm_mul_le_const_mul_norm {i : ι} (hBi : B i = (1 : L))
+      (hna : IsNonarchimedean (Norm.norm : K → ℝ)) :
+    ∃ (c : ℝ) (_ : 0 < c), ∀ x y : L, B.norm (x * y) ≤ c * B.norm x * B.norm y := by
+  -- The bounding constant `c` will be the maximum of the products `B.norm (B i * B j)`.
+  obtain ⟨c, _, hc⟩ := exists_mem_eq_sup' univ_nonempty (fun i : ι × ι ↦ B.norm (B i.1 * B i.2))
+  use B.norm (B c.1 * B c.2)
+  constructor
+  -- ∀ (x y : L), B.norm (x * y) ≤ B.norm (⇑B c.fst * ⇑B c.snd) * B.norm x * B.norm y
+  · intro x y
+    -- `ixy` is an index for which `‖B.repr (x*y) i‖` is maximum.
+    obtain ⟨ixy, _, hixy_def⟩ := exists_mem_eq_sup' univ_nonempty (fun i ↦ ‖(B.repr (x * y)) i‖)
+    -- We rewrite the LHS using `ixy`.
+    conv_lhs => simp only [Basis.norm]; rw [hixy_def, ← Basis.sum_repr B x, ← Basis.sum_repr B y]
+    rw [sum_mul, map_finset_sum]
+    simp_rw [smul_mul_assoc, LinearEquiv.map_smul, mul_sum, map_finset_sum,
+      mul_smul_comm, LinearEquiv.map_smul]
+    have hna' : IsNonarchimedean (NormedField.toMulRingNorm K) := hna
+    /- Since the norm is nonarchimidean, the norm of a finite sum is bounded by the maximum of the
+          norms of the summands. -/
+    have hk : ∃ (k : ι) (_ : (univ : Finset ι).Nonempty → k ∈ univ ),
+        ‖∑ i : ι, (B.repr x i • ∑ i_1 : ι, B.repr y i_1 • B.repr (B i * B i_1)) ixy‖ ≤
+          ‖(B.repr x k • ∑ j : ι, B.repr y j • B.repr (B k * B j)) ixy‖ := by
+      obtain ⟨k, hk0, hk⟩ := IsNonarchimedean.finset_image_add hna'
+        (fun i ↦ (B.repr x i • ∑ i_1 : ι, B.repr y i_1 • B.repr (B i * B i_1)) ixy)
+        (univ : Finset ι)
+      exact ⟨k, hk0, hk⟩
+    obtain ⟨k, -, hk⟩ := hk
+    simp only [Finsupp.coe_smul, Finsupp.coe_finset_sum, Pi.smul_apply, sum_apply, smul_eq_mul,
+      norm_mul, sup'_le_iff, mem_univ, forall_const] at hk ⊢
+    apply le_trans hk
+    -- We use the above property again.
+    have hk' : ∃ (k' : ι) (_ : (univ : Finset ι).Nonempty → k' ∈ univ),
```

Comment:
Similarly you can just do `obtain ⟨k', hk'⟩` here (and drop the `(_ : (univ : Finset ι).Nonempty → k' ∈ univ),`) rather than making the first `hk'` and then destructing it on line 134 (which you can now delete). I don't think we lose readability here, this is one thing I like about the `obtain` tactic.
```suggestion
    obtain ⟨k', hk'⟩ : ∃ (k' : ι),
```